package/src/hooks/gtm/index.tsx

```diff
@@ -0,0 +1,62 @@
+import React, {
+  ReactNode,
+  createContext,
+  useEffect,
+  useContext,
+  useReducer,
+} from "react";
+import { IGtmParams, initGTM, sendToGTM } from "./GoogleTagManager";
+import * as constants from "../../ApiContext/api/constants";
+
+export type GTMProviderProps = {
+  state?: IGtmParams;
+  children: ReactNode;
+};
+
+const initialState: IGtmParams = {
+  gtmId: "",
+  dataLayer: undefined,
+  dataLayerName: constants.DEFAULT_GTM_DATA_LAYER,
+};
+
+export const GTMContext = createContext<IGtmParams | undefined>(initialState);
+export const GTMContextDispatch = createContext<any | undefined>(undefined);
+
+function dataReducer(state: IGtmParams, data: any) {
+  sendToGTM({ dataLayerName: state?.dataLayer?.dataLayerName!, data });
+  return state;
+}
+
+const GTMProvider: React.FC<GTMProviderProps> = ({ state, children }) => {
+  const [store, dispatch] = useReducer(dataReducer, {
+    ...initialState,
+    ...state,
+  });
+
+  const mergedState = { ...store, ...state };
+  useEffect(() => {
+    initGTM({
+      gtmId: mergedState.gtmId,
+      dataLayer: mergedState.dataLayer,
+      dataLayerName: mergedState.dataLayerName,
+    });
+  }, [state]);
+
+  return (
+    <GTMContext.Provider value={store}>
+      <GTMContextDispatch.Provider value={dispatch}>
+        {children}
+      </GTMContextDispatch.Provider>
+    </GTMContext.Provider>
+  );
+};
+
+function useGTMDispatch() {
+  const context = useContext(GTMContextDispatch);
+  if (context === undefined)
+    throw new Error("dispatchGTMEvent must be used within a GTMProvider");
```

Comment:
Lets have the errror messages in a common class and have here.

Reply:
@nyshthefantastic this hook is written as an external lib. That's why I kept it here. This error comes in developer level when a developer uses this tag out side the defined context while developing. This is not a widget related error. Widget users won't see this error. There are other application related errors in the code base we should move to a common class. We will move those eventually while refactoring the code.